bufferevent_mbedtls.c

```diff
@@ -0,0 +1,1379 @@
+	} else {
+		struct bufferevent *bev = &bev_ssl->bev.bev;
+		event_del(&bev->ev_write);
+	}
+}
+
+static int
+set_rbow(struct bufferevent_mbedtls *bev_ssl)
+{
+	if (!bev_ssl->underlying)
+		stop_reading(bev_ssl);
+	bev_ssl->read_blocked_on_write = 1;
+	return start_writing(bev_ssl);
+}
+
+static int
+set_wbor(struct bufferevent_mbedtls *bev_ssl)
+{
+	if (!bev_ssl->underlying)
+		stop_writing(bev_ssl);
+	bev_ssl->write_blocked_on_read = 1;
+	return start_reading(bev_ssl);
+}
+
+static int
+clear_rbow(struct bufferevent_mbedtls *bev_ssl)
+{
+	struct bufferevent *bev = &bev_ssl->bev.bev;
+	int r = 0;
+	bev_ssl->read_blocked_on_write = 0;
+	if (!(bev->enabled & EV_WRITE))
+		stop_writing(bev_ssl);
+	if (bev->enabled & EV_READ)
+		r = start_reading(bev_ssl);
+	return r;
+}
+
+
+static int
+clear_wbor(struct bufferevent_mbedtls *bev_ssl)
+{
+	struct bufferevent *bev = &bev_ssl->bev.bev;
+	int r = 0;
+	bev_ssl->write_blocked_on_read = 0;
+	if (!(bev->enabled & EV_READ))
+		stop_reading(bev_ssl);
+	if (bev->enabled & EV_WRITE)
+		r = start_writing(bev_ssl);
+	return r;
+}
+
+static void
+conn_closed(struct bufferevent_mbedtls *bev_ssl, int when, int errcode, int ret)
+{
+	int event = BEV_EVENT_ERROR;
+	//int dirty_shutdown = 0;
+	char buf[100];
+
+	if (when & BEV_EVENT_READING && ret == 0)
+	{ 
+		if (bev_ssl->allow_dirty_shutdown)
+			event = BEV_EVENT_EOF;
+	} else {
+		mbedtls_strerror(errcode, buf, sizeof(buf));
+		switch (errcode) {
+		case MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY:
+			event = BEV_EVENT_EOF;
+			break;
+		//case MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS:
+		//case MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS:
+		case MBEDTLS_ERR_SSL_CLIENT_RECONNECT:
+			event_warnx("BUG: Unsupported feature %d: %s", errcode, buf);
+			break;
+		default:
+			/* should be impossible; treat as normal error. */
+			event_warnx("BUG: Unexpected mbedtls error code %d: %s", errcode, buf);
+			break;
+		}
+
+		put_error(bev_ssl, errcode);
+	}
+
+
+	stop_reading(bev_ssl);
+	stop_writing(bev_ssl);
+
+	bufferevent_run_eventcb_(&bev_ssl->bev.bev, when | event, 0);
+}
+
+/*static void
+init_bio_counts(struct bufferevent_mbedtls *bev_ssl)
+{
+	BIO *rbio, *wbio;
+
+	wbio = SSL_get_wbio(bev_ssl->ssl);
+	bev_ssl->counts.n_written = wbio ? BIO_number_written(wbio) : 0;
+	rbio = SSL_get_rbio(bev_ssl->ssl);
+	bev_ssl->counts.n_read = rbio ? BIO_number_read(rbio) : 0;
+}
+
+static inline void
+decrement_buckets(struct bufferevent_mbedtls *bev_ssl)
+{
+	unsigned long num_w = BIO_number_written(SSL_get_wbio(bev_ssl->ssl));
+	unsigned long num_r = BIO_number_read(SSL_get_rbio(bev_ssl->ssl));
+	/ * These next two subtractions can wrap around. That's okay. * /
+	unsigned long w = num_w - bev_ssl->counts.n_written;
+	unsigned long r = num_r - bev_ssl->counts.n_read;
+	if (w)
+		bufferevent_decrement_write_buckets_(&bev_ssl->bev, w);
+	if (r)
+		bufferevent_decrement_read_buckets_(&bev_ssl->bev, r);
+	bev_ssl->counts.n_written = num_w;
+	bev_ssl->counts.n_read = num_r;
+}*/
```

Comment:
it is worrying to see all this commented-out code left in ... it seems this is not relevant for mbedtls, and should be cleaned up?